File: create_airtable_form_api.js
#!/usr/bin/env node
/**
 * Create Airtable Form using API
 * This script creates a complete form setup in Airtable for the iME questionnaire
 */

const fetch = require('node-fetch');
require('dotenv').config();

// Load form configuration
const formConfig = require('./data/airtable_form_config.json');

class AirtableFormCreator {
    constructor() {
        this.apiKey = process.env.AIRTABLE_API_KEY;
        this.baseId = process.env.AIRTABLE_BASE_ID;
        this.headers = {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json'
        };
    }

    /**
     * Create a new table with all form fields
     */
    async createFormTable() {
        console.log('Creating Form Responses table...');
        
        // Build field definitions
        const fields = [
            {
                name: 'Response ID',
                type: 'autoNumber'
            },
            {
                name: 'Company Name',
                type: 'singleLineText'
            },
            {
                name: 'Email',
                type: 'email'
            },
            {
                name: 'Submission Date',
                type: 'dateTime',
                options: {
                    timeZone: 'America/New_York',
                    dateFormat: {
                        name: 'us'
                    }
                }
            }
        ];

        // Add all questionnaire fields
        formConfig.sections.forEach(section => {
            section.fields.forEach(field => {
                const fieldDef = this.createFieldDefinition(field);
                fields.push(fieldDef);
            });
        });

        // Create table
        const createTableUrl = `https://api.airtable.com/v0/meta/bases/${this.baseId}/tables`;
        
        try {
            const response = await fetch(createTableUrl, {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify({
                    name: 'Form Responses',
                    fields: fields
                })
            });

            if (!response.ok) {
                const error = await response.text();
                throw new Error(`Failed to create table: ${error}`);
            }

            const result = await response.json();
            console.log('✓ Table created successfully');
            return result.id;
        } catch (error) {
            console.error('Error creating table:', error);
            throw error;
        }
    }

    /**
     * Convert field configuration to Airtable field definition
     */
    createFieldDefinition(field) {
        const fieldDef = {
            name: field.field_name,
            description: field.field_description || ''
        };

        switch (field.field_type) {
            case 'single_select':
                fieldDef.type = 'singleSelect';
                fieldDef.options = {
                    choices: field.options.map(opt => ({
                        name: opt.label,
                        color: this.getColorForOption(opt.value)
                    }))
                };
                break;

            case 'multiple_select':
                fieldDef.type = 'multipleSelects';
                fieldDef.options = {
                    choices: field.options.map(opt => ({
                        name: opt.label,
                        color: this.getColorForOption(opt.value)
                    }))
                };
                break;

            case 'long_text':
                fieldDef.type = 'multilineText';
                break;

            default:
                fieldDef.type = 'singleLineText';
        }

        return fieldDef;
    }

    /**
     * Assign colors to options for better visual organization
     */
    getColorForOption(value) {
        const colorMap = {
            // Technical areas - blues/purples
            'ai_ml': 'blueBright',
            'cybersecurity': 'purpleBright',
            'autonomy': 'blueDark',
            'communications': 'cyanBright',
            'sensors': 'tealBright',
            
            // Agencies - greens
            'army': 'greenBright',
            'navy': 'greenDark',
            'air_force': 'greenLight',
            'space_force': 'tealDark',
            
            // Risk levels - yellows/oranges/reds
            'conservative': 'yellowBright',
            'moderate': 'orangeBright',
            'aggressive': 'redBright',
            
            // Default
            'default': 'grayLight'
        };

        return colorMap[value] || colorMap['default'];
    }

    /**
     * Create a form view for the table
     */
    async createFormView(tableId) {
        console.log('Creating form view...');
        
        const viewUrl = `https://api.airtable.com/v0/meta/bases/${this.baseId}/tables/${tableId}/views`;
        
        // Build visible fields list with descriptions
        const visibleFields = ['Company Name', 'Email'];
        formConfig.sections.forEach(section => {
            section.fields.forEach(field => {
                visibleFields.push(field.field_name);
            });
        });

        const viewConfig = {
            name: 'iME Onboarding Form',
            type: 'form',
            metadata: {
                viewSectionConfigs: this.createFormSections()
            }
        };

        try {
            const response = await fetch(viewUrl, {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify(viewConfig)
            });

            if (!response.ok) {
                const error = await response.text();
                throw new Error(`Failed to create view: ${error}`);
            }

            const result = await response.json();
            console.log('✓ Form view created successfully');
            return result;
        } catch (error) {
            console.error('Error creating form view:', error);
            throw error;
        }
    }

    /**
     * Create form sections configuration
     */
    createFormSections() {
        const sections = [{
            name: 'Company Information',
            fields: ['Company Name', 'Email']
        }];

        formConfig.sections.forEach(section => {
            sections.push({
                name: section.section_title,
                description: section.section_description,
                fields: section.fields.map(f => f.field_name)
            });
        });

        return sections;
    }

    /**
     * Update field properties (required, descriptions, etc)
     */
    async updateFieldProperties(tableId) {
        console.log('Updating field properties...');
        
        // Get current fields
        const tableUrl = `https://api.airtable.com/v0/meta/bases/${this.baseId}/tables/${tableId}`;
        const response = await fetch(tableUrl, { headers: this.headers });
        const table = await response.json();
        
        // Update each field with required status and descriptions
        for (const field of table.fields) {
            const fieldConfig = this.findFieldConfig(field.name);
            if (!fieldConfig) continue;

            if (fieldConfig.is_required || field.name === 'Company Name' || field.name === 'Email') {
                // Note: Airtable API doesn't directly support setting required fields
                // This would need to be done manually or through the form configuration
                console.log(`  - ${field.name}: Mark as required in form settings`);
            }

            if (fieldConfig.validation?.max_selections) {
                console.log(`  - ${field.name}: Set max selections to ${fieldConfig.validation.max_selections}`);
            }
        }
    }

    /**
     * Find field configuration by name
     */
    findFieldConfig(fieldName) {
        for (const section of formConfig.sections) {
            const field = section.fields.find(f => f.field_name === fieldName);
            if (field) return field;
        }
        return null;
    }

    /**
     * Generate a shareable form link
     */
    generateFormLink(viewId) {
        return `https://airtable.com/shr${viewId}`;
    }

    /**
     * Main execution
     */
    async createForm() {
        try {
            console.log('=== Airtable Form Creator ===\n');
            
            // Step 1: Create table
            const tableId = await this.createFormTable();
            
            // Step 2: Create form view
            const view = await this.createFormView(tableId);
            
            // Step 3: Update field properties
            await this.updateFieldProperties(tableId);
            
            // Summary
            console.log('\n=== Setup Complete! ===');
            console.log(`Table ID: ${tableId}`);
            console.log(`View ID: ${view.id}`);
            console.log(`Form URL: ${this.generateFormLink(view.id)}`);
            
            console.log('\n=== Next Steps ===');
            console.log('1. Open Airtable and navigate to your base');
            console.log('2. Find the "Form Responses" table');
            console.log('3. Click on the "iME Onboarding Form" view');
            console.log('4. Configure these form settings:');
            console.log('   - Mark required fields');
            console.log('   - Set up conditional logic');
            console.log('   - Customize form design and branding');
            console.log('   - Configure email notifications');
            console.log('5. Test the form with sample data');
            
            // Save configuration
            const config = {
                tableId,
                viewId: view.id,
                formUrl: this.generateFormLink(view.id),
                createdAt: new Date().toISOString()
            };
            
            const fs = require('fs');
            fs.writeFileSync('data/airtable_form_config_result.json', JSON.stringify(config, null, 2));
            console.log('\nConfiguration saved to: data/airtable_form_config_result.json');
            
        } catch (error) {
            console.error('Failed to create form:', error);
            process.exit(1);
        }
    }
}

// Alternative: Script for Airtable's Scripting Block
function generateAirtableScript() {
    console.log('\n=== Alternative: Airtable Scripting Block Code ===\n');
    
    const script = `
// Airtable Scripting Block Code
// Run this inside your Airtable base

// Create all fields programmatically
const table = base.getTable('Form Responses');

// Field configurations from our form config
const fieldConfigs = ${JSON.stringify(formConfig.sections.flatMap(s => s.fields), null, 2)};

// Create records for testing
const testRecord = {
    'Company Name': 'Test Company Inc.',
    'Email': 'test@example.com'
};

// Add questionnaire responses
fieldConfigs.forEach(field => {
    if (field.field_type === 'single_select' && field.options.length > 0) {
        testRecord[field.field_name] = field.options[0].label;
    } else if (field.field_type === 'multiple_select' && field.options.length > 0) {
        testRecord[field.field_name] = [field.options[0].label];
    } else if (field.field_type === 'long_text') {
        testRecord[field.field_name] = 'Sample text response';
    }
});

// Create test record
await table.createRecordAsync(testRecord);
console.log('Test record created successfully!');
`;

    console.log(script);
}

// Run the script
if (require.main === module) {
    const creator = new AirtableFormCreator();
    
    // Check for required environment variables
    if (!process.env.AIRTABLE_API_KEY || !process.env.AIRTABLE_BASE_ID) {
        console.error('Error: Missing required environment variables');
        console.error('Please set AIRTABLE_API_KEY and AIRTABLE_BASE_ID in your .env file');
        
        // Generate scripting block code as alternative
        generateAirtableScript();
        process.exit(1);
    }
    
    // Create the form
    creator.createForm().catch(console.error);
}

module.exports = AirtableFormCreator;

File: create_airtable_form_api.py
#!/usr/bin/env python3
"""
Create Airtable Form using API
This script creates a complete form setup in Airtable for the iME questionnaire
"""

import os
import json
import requests
from typing import Dict, List, Any
from dotenv import load_dotenv

load_dotenv()

class AirtableFormCreator:
    def __init__(self):
        self.api_key = os.getenv('AIRTABLE_API_KEY')
        self.base_id = os.getenv('AIRTABLE_BASE_ID')
        self.headers = {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json'
        }
        
        # Load form configuration
        with open('data/airtable_form_config.json', 'r') as f:
            self.form_config = json.load(f)
    
    def create_form_table(self) -> str:
        """Create a new table with all form fields"""
        print('Creating Form Responses table...')
        
        # Build field definitions
        fields = [
            {
                'name': 'Response ID',
                'type': 'autoNumber'
            },
            {
                'name': 'Company Name',
                'type': 'singleLineText'
            },
            {
                'name': 'Email',
                'type': 'email'
            },
            {
                'name': 'Submission Date',
                'type': 'dateTime',
                'options': {
                    'timeZone': 'America/New_York',
                    'dateFormat': {'name': 'us'}
                }
            }
        ]
        
        # Add all questionnaire fields
        for section in self.form_config['sections']:
            for field in section['fields']:
                field_def = self.create_field_definition(field)
                fields.append(field_def)
        
        # Create table via API
        url = f'https://api.airtable.com/v0/meta/bases/{self.base_id}/tables'
        
        response = requests.post(url, headers=self.headers, json={
            'name': 'Form Responses',
            'fields': fields
        })
        
        if response.status_code == 200:
            result = response.json()
            print('✓ Table created successfully')
            return result['id']
        else:
            print(f'✗ Failed to create table: {response.text}')
            raise Exception(f'API Error: {response.status_code}')
    
    def create_field_definition(self, field: Dict) -> Dict:
        """Convert field configuration to Airtable field definition"""
        field_def = {
            'name': field['field_name'],
            'description': field.get('field_description', '')
        }
        
        if field['field_type'] == 'single_select':
            field_def['type'] = 'singleSelect'
            field_def['options'] = {
                'choices': [
                    {'name': opt['label']} 
                    for opt in field.get('options', [])
                ]
            }
        elif field['field_type'] == 'multiple_select':
            field_def['type'] = 'multipleSelects'
            field_def['options'] = {
                'choices': [
                    {'name': opt['label']} 
                    for opt in field.get('options', [])
                ]
            }
        elif field['field_type'] == 'long_text':
            field_def['type'] = 'multilineText'
        else:
            field_def['type'] = 'singleLineText'
        
        return field_def
    
    def generate_manual_script(self):
        """Generate a manual setup script for Airtable"""
        print("\n=== Manual Setup Script ===")
        print("Copy and paste this into Airtable's Scripting app:\n")
        
        script = """// Airtable Manual Form Setup Script
// This script helps you configure your form manually

// 1. First, create these fields in your table:
const requiredFields = [
    'Company Name (Single line text) - Required',
    'Email (Email) - Required',
"""
        
        # Add all fields
        for section in self.form_config['sections']:
            script += f"\n    // {section['section_title']}\n"
            for field in section['fields']:
                field_type = field['field_type'].replace('_', ' ').title()
                required = "Required" if field['is_required'] else "Optional"
                script += f"    '{field['field_name']} ({field_type}) - {required}',\n"
        
        script += """];

// 2. Configure form view settings:
const formSettings = {
    title: 'iME Business Profile Questionnaire',
    description: 'Help us match you with the best defense contracting opportunities',
    submitButtonText: 'Submit Profile',
    redirectUrl: '', // Add your thank you page URL
    coverImage: '', // Add your branding image URL
};

// 3. Field-specific settings:
"""
        
        # Add field configurations
        for section in self.form_config['sections']:
            script += f"\n// {section['section_title']}\n"
            for field in section['fields']:
                script += f"// {field['field_name']}:\n"
                if field.get('field_description'):
                    script += f"//   - Description: {field['field_description']}\n"
                if field.get('validation', {}).get('max_selections'):
                    script += f"//   - Max selections: {field['validation']['max_selections']}\n"
                if field.get('options'):
                    script += f"//   - Options: {len(field['options'])} choices\n"
                script += "\n"
        
        script += """
// 4. After creating all fields:
//    a) Go to Forms view
//    b) Add all fields in the order shown above
//    c) Mark required fields
//    d) Add field descriptions
//    e) Configure conditional logic if needed

output.text('Setup instructions generated! Follow the steps above.');
"""
        
        print(script)
        
        # Save to file
        with open('data/airtable_manual_setup.js', 'w') as f:
            f.write(script)
        print("\nScript saved to: data/airtable_manual_setup.js")
    
    def generate_field_summary(self):
        """Generate a summary of all fields for easy reference"""
        print("\n=== Field Summary ===")
        
        total_fields = 4  # Base fields
        required_fields = 2  # Company Name, Email
        
        for section in self.form_config['sections']:
            print(f"\n{section['section_title']}:")
            for field in section['fields']:
                total_fields += 1
                if field['is_required']:
                    required_fields += 1
                
                field_type = field['field_type'].replace('_', ' ')
                req = "Required" if field['is_required'] else "Optional"
                print(f"  - {field['field_name']} ({field_type}) [{req}]")
                
                if field.get('field_description'):
                    print(f"    Helper: {field['field_description']}")
                if field.get('options'):
                    print(f"    Options: {len(field['options'])} choices")
                if field.get('validation', {}).get('max_selections'):
                    print(f"    Max selections: {field['validation']['max_selections']}")
        
        print(f"\nTotal fields: {total_fields}")
        print(f"Required fields: {required_fields}")
        print(f"Optional fields: {total_fields - required_fields}")

def main():
    """Main execution"""
    creator = AirtableFormCreator()
    
    print("=== Airtable Form Creator ===\n")
    
    # Check for API credentials
    if not creator.api_key or not creator.base_id:
        print("✗ Missing Airtable credentials!")
        print("\nTo use the API method, add to your .env file:")
        print("AIRTABLE_API_KEY=your_api_key")
        print("AIRTABLE_BASE_ID=your_base_id")
        print("\nGenerating manual setup script instead...")
        
        creator.generate_manual_script()
        creator.generate_field_summary()
        
        print("\n=== Next Steps ===")
        print("1. Open your Airtable base")
        print("2. Create a new table called 'Form Responses'")
        print("3. Add all fields listed in the script")
        print("4. Create a Form view")
        print("5. Configure form settings and field order")
        return
    
    # Try to create via API
    try:
        table_id = creator.create_form_table()
        
        print(f"\n✓ Success! Table created with ID: {table_id}")
        print("\n=== Next Steps ===")
        print("1. Open your Airtable base")
        print("2. Find the 'Form Responses' table")
        print("3. Create a Form view")
        print("4. Configure these settings:")
        print("   - Mark required fields")
        print("   - Add field descriptions from the config")
        print("   - Set up conditional logic")
        print("   - Customize form branding")
        
        # Save result
        result = {
            'table_id': table_id,
            'base_id': creator.base_id,
            'created_at': str(datetime.now())
        }
        
        with open('data/airtable_form_result.json', 'w') as f:
            json.dump(result, f, indent=2)
        print("\nConfiguration saved to: data/airtable_form_result.json")
        
    except Exception as e:
        print(f"\n✗ API creation failed: {e}")
        print("\nGenerating manual setup script instead...")
        creator.generate_manual_script()
        creator.generate_field_summary()

if __name__ == "__main__":
    from datetime import datetime
    main()